### PR TITLE
feat(compression): implement tensor decompression in op concatenation

### PR DESCRIPTION
**Pull request by rkuester**
Implement tensor decompression in op concatenation. Extend
tests to validate operation on compressed tensors.

BUG=part of #2636

### CODE REVIEW

File: tensorflow/lite/micro/kernels/concatenation.cc [hunk 1]
```diff
@@ -155,62 +176,53 @@ TfLiteStatus ConcatenationPrepare(TfLiteContext* context, TfLiteNode* node) {
...
+    if (input_type == kTfLiteInt8) {
+      // Make sure there is no re-scaling needed for Int8 quantized kernel. This
+      // is a restriction we introduced to Int8 kernels.
+      TF_LITE_ENSURE_EQ(context, static_cast<double>(input->params.scale),
+                        static_cast<double>(output_tensor->params.scale));
+      TF_LITE_ENSURE_EQ(context, input->params.zero_point,
+                        output_tensor->params.zero_point);
+    } else if (input_type == kTfLiteInt16) {
+      // Make sure that all Int16 inputs have a null zero-point.
+      TF_LITE_ENSURE_EQ(context, input->params.zero_point, 0);
```
Comment (yushhuang) [inline tensorflow/lite/micro/kernels/concatenation.cc:189]: Is there a reason why the scale is not checked in the int16 case? AFAICS the (at least the reference) concatenation kernel does not take into account any scaling/quantization at all.

https://github.com/tensorflow/tflite-micro/blob/b2f27186d243d6afc2324c43b3bf9a9f681be474/tensorflow/lite/kernels/internal/reference/concatenation.h#L30

**Review by yushhuang — APPROVED** [review]

File: tensorflow/lite/micro/kernels/concatenation.cc (hunk 1 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/concatenation.cc:189]: I agree with your conclusion, however this is a copy-paste of the TfLite code.

File: tensorflow/lite/micro/kernels/concatenation.cc [hunk 2]
```diff
@@ -155,62 +176,53 @@ TfLiteStatus ConcatenationPrepare(TfLiteContext* context, TfLiteNode* node) {
...
-        TF_LITE_ENSURE(context, t != nullptr);
-        input_scales[i] = t->params.scale;
-        input_zero_points[i] = t->params.zero_point;
-        micro_context->DeallocateTempTfLiteTensor(t);
-      }
-
-      data->params.input_scale = input_scales;
-      data->params.input_zeropoint = input_zero_points;
-      data->params.output_zeropoint = output->params.zero_point;
-      data->params.output_scale = output->params.scale;
```
Comment (suleshahid) [inline tensorflow/lite/micro/kernels/concatenation.cc:204]: Why do we not need to set these anymore?

Comment (suleshahid) [reply tensorflow/lite/micro/kernels/concatenation.cc:204]: Never mind, I followed the code, yea we don't use this, only for the other ConcatWithScaling which TFLM doesn't use.

**Review by suleshahid — APPROVED** [review]

File: tensorflow/lite/micro/kernels/concatenation.cc (hunk 1 above)
Comment (tinskip) [reply tensorflow/lite/micro/kernels/concatenation.cc:189]: Yes, also is there are reason why it has to be zero? Seems that concatenation in general shouldn't care about quantization parameters, as long as they match in and out.

Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/concatenation.cc:189]: @tinskip This same check is performed by the LiteRT (TfLite) reference implementation.

**Pull request merged**